### DISCUSSION TITLE
Accesing vitest Globals

### DISCUSSION THREAD

**Discussion by odinvanderlinden** [vitest-dev/vitest#3017] [Q&A]
Currently we are using files to communicate between the vitest VM and our main process for the vitest plugin in stryker.
As mentioned in #3015
> The public API is not standardised. We can extend it based on the feedback if needed.

We were wondering if there could be added a way to access the global variables from the VM. For instance like this:
```ts
this.ctx = await createVitest('test', {});
this.ctx.globals.__stryker__ = {}
```

**Comment by sheremet-va** [comment]
`createVitest` is running in the main thread. Tests are running in `child_process` or `worker_thread`, you cannot access objects between them directly.

**Reply by nicojs** [reply]
Interesting. How are you communicating the coverage object in the case of `instanbul` or `c8`?

**Reply by sheremet-va** [reply]
Via rpc. When tests are finished running, we send the object to the main thread:

https://github.com/vitest-dev/vitest/blob/d4d04257c0c47f0acd7e508fbac74d0a57330a58/packages/vitest/src/runtime/entry.ts#L60

**Reply by nicojs** [reply]
That seems great to me. Are you open to opening some of that on the public API?

```ts
const ctx = await createVitest('test', {});
ctx.setGlobals({ __stryker__: { hitCounter: 0, activeMutant: "43" } });
ctx.globalRecovery = ['__stryker__'];
const result = await ctx.start();
console.log(result.global.__stryker__);
```

**Reply by sheremet-va** [reply]
I'm unsure how good it is to expose such things in the Vitest context. `rpc` is unique for each thread. You can see that coverage collects this data and doesn't just use it: 

https://github.com/vitest-dev/vitest/blob/3573032875eac78eec55f37c7d42505bf2f0d4cb/packages/coverage-c8/src/provider.ts#L65

You need to have your own mechanism to handle this which is unique to your case. But for example we can expose a way to extend `rpc` with your own methods, and add `rpc` instance to the [runner](https://vitest.dev/advanced/runner.html) just so it's easier to reuse.

**Reply by nicojs** [reply]
Ah yes, that makes sense. We're doing all of that now via files we write from each worker and read from the host process. Using `rpc` might yield a performance benefit, but [hot-reload](https://stryker-mutator.io/blog/stryker-js-v6-expeditious-superior-mutations/#-hot-reload) would probably be an order of magnitude performance improvement. I'll open a separate discussion for that once we've implemented the first version of the vitest plugin.

**Comment by nicojs** [comment]
@sheremet-va I've noticed that `vitest.state` is shared between runs.

```ts
const vitest = await createVitest("test", {
  onConsoleLog: () => false,
  watch: false,
  threads: false,
});
await vitest.start(['packages/bar/src/bar-math.spec.js']);
const tests = collectTests(); // collects tests from the vitest.state
console.log(tests); // 1 test
await vitest.start(['packages/foo/src/foo-math.spec.js']);
console.log(collectTests()); // 2 tests, both bar-math.spec and foo-math.spec
```

What is the correct way to clean state between runs? My best guess is this:

```ts
function reset() {
  vitest.state.getFiles().forEach(file => delete file.suite);
  vitest.state.clearErrors();
}
```

<details>
<summary><code>collectTests</code></summary>

```ts

function collectTests() {
  return vitest.state.getFiles().flatMap(collectTestsFromSuite);
}

/**
 *
 * @param {import('vitest').Suite} suite
 * @param {import('vitest').Vitest['projects'][0]}
 * @returns {string[]}
 */
function collectTestsFromSuite(suite) {
  return suite.tasks.flatMap((task) => {
    if (task.type === "suite") {
      return collectTestsFromSuite(task);
    } else if (task.type === "test") {
      return toTestId(task);
    } else {
      return [];
    }
  });
}

/**
 *
 * @param {import('vitest').Test} test
 * @returns {string}
 */
function toTestId(test) {
  function collectTestName({ name, suite }) {
    const nameParts = [name];
    let currentSuite = suite;
    while (currentSuite) {
      nameParts.unshift(currentSuite.name);
      currentSuite = currentSuite.suite;
    }
    return nameParts.join(" ").trim();
  }
  return `${test.file?.filepath ?? ""}#${collectTestName(test)}\n\t${
    test.result?.state
  }`;
}
```

</details>

**Reply by sheremet-va** [reply]
What is `collectTests`? Tests are cleared between runs.

**Reply by sheremet-va** [reply]
> What is collectTests?

Oh, I just noticed a spoiler 🤦🏻

**Reply by sheremet-va** [reply]
I guess if you want to clear it, `state.filesMap.clear()` should do the trick. `vitest.state.clearErrors` is already called inside `start` function.

**Reply by nicojs** [reply]
> What is `collectTests`? Tests are cleared between runs.

[run-vitest.zip](https://github.com/vitest-dev/vitest/files/11474882/run-vitest.zip)

I don't think they are. At least not if you switch filters between runs. See my zip file for an example project.

**Reply by nicojs** [reply]
> I guess if you want to clear it, `state.filesMap.clear()` should do the trick. `vitest.state.clearErrors` is already called inside `start` function.

Ah yes, that does the trick! Thanks!

F.y.i. do you want us to contribute tests and docs for this and other API's we're using (i.e. `setupFiles`, etc)? Since Stryker will rely on this, it is in our best interest it keeps being supported, or if not that we get a notification :)

**Reply by sheremet-va** [reply]
Yes, I checked and we only clear tests that are running (`packages/foo/src/foo-math.spec.js` would be cleared if it was already inside a state).

**Reply by sheremet-va** [reply]
> F.y.i. do you want us to contribute tests and docs for this and other API's we're using (i.e. setupFiles, etc)? Since Stryker will rely on this, it is in our best interest it keeps being supported, or if not that we get a notification :)

Feel free to contribute to the public API page 😄

**Comment by nicojs** [comment]
@sheremet-va our current solution we have makes use of `fs.writeFile` and `fs.readFile` and has several drawbacks.
- doesn't work if a user mocks away `fs.writeFile`
- it is not great for performance
- most importantly: it doesn't work with browser mode.

Having some experience now with vitests design, i think we should be able to implement something like this:

```js
afterAll((suite) => {
  suite.result.custom.strykerData = stuff;
});
```

Do you see problems with such an approach? Would you be open for this change?

**Reply by sheremet-va** [reply]
> Having some experience now with vitests design, i think we should be able to implement something like this:

And exactly how would you use it?

----

I propose this API for communicating between server and runtime:

```ts
// inside the runtime
import { rpc } from 'vitest/rpc'

onSomething(() => {
  rpc().updateStryker(stuff)
})
```

```ts
// main config (not workspace project)
export default {
  test: {
    rpc: {
      extendMain: './path/to/node-module.js',
      extendRuntime: './path/to/runtime-module.js'
    }
  }
}
```

```ts
// path/to/node-module.js
// this is imported inside node and can use any node APi

// handlers are triggered from runtime and sent to the node process
export function handlers(ctx: Vitest, workspaceProject: WorkspaceProject) {
  return {
    updateStryker(stuff) {
      // do stuff
    }
  }
}

// events are triggered from node process and sent to the runtime
export function events(ctx: Vitest, workspaceProject: WorkspaceProject) {
  return ['onSendData']
}

export function afterCreate(ctx: Vitest, workspaceProject: WorkspaceProject, rpc) {
  // if you send any events from the node process, you can send them to the runtime from here
  vitest.stryker.onSomethingImportant(data => rpc.onSendData(data))
}
```

```ts
// path/to/runtime-module.js
// this should not rely on anything node-related

// register events from `events` function in node-module
export function events() {
  return {
   onSomethingImportant(data) {
     // do something important!
   }
  }
}
```

Just to let you know, RPC is created per test file if tests are isolated. If they are not isolated RPC is shared between tests inside the same `environment`. So I don't recommend storing it anywhere.

Another thing is that `data` is sent differently depending on the environment. If tests are running inside a worker, they are cloned with `structuredClone`. If tests are running inside `child_process`, they are serialized and parsed with v8.serialize/v8.deserialize. And browser RPC uses web sockets for communicating with `flatted` module to send/receive data, so it can still have cyclic references.

**Reply by nicojs** [reply]
> > Having some experience now with vitests design, i think we should be able to implement something like this:
> 
> And exactly how would you use it?

Ah, yes, sorry I wasn't clear about that part 😅. It would look something like this:

```js
const vitest = await createVitest("test", { watch: false });
vitest.projects[0].config.setupFiles = ["vitest.setup.js"];
await vitest.start();
vitest.state.getFiles()[0].result.custom.strykerData; // per suite/worker we can get the custom data here
```

Anything on `custom` (or `metadata` or something?) is serialized back to the main process.

To be clear, all we need is a way to communicate from the worker to the main node process after the worker is completed (`worker --> main` ). The other way around (`main --> worker`) is covered using a custom `vitest.setup.js` file (registered as inline dep and flushed from the cache between runs).

How you describe using `rpc` probably works to, but it seems pretty complicated and has a big API surface. I am also not sure how it would integrate with `afterAll` (would `rpc().updateStryker(stuff)` be `await`-able?)

If you are interested in our current setup, you can see it here: https://github.com/stryker-mutator/stryker-js/blob/master/packages/vitest-runner/src/vitest-test-runner.ts. We will release the plugin this Friday I hope 🤞. This first implementation uses files to communicate between processes, so only the node-based runner environment is supported.

**Reply by sheremet-va** [reply]
> How you describe using rpc probably works to, but it seems pretty complicated and has a big API surface. I am also not sure how it would integrate with afterAll (would rpc().updateStryker(stuff) be await-able?)

Yes, they are awaitable. This is how Vitest communicates with runtime already.

> Anything on custom (or metadata or something?) is serialized back to the main process.

There is already a `meta` property actually on the suite that is meant for libraries, but I think it's currently bugged and is not sent to the server (there is even an issue somewhere...)

**Reply by nicojs** [reply]
> > Anything on custom (or metadata or something?) is serialized back to the main process.
> 
> There is already a `meta` property actually on the suite that is meant for libraries, but I think it's currently bugged and is not sent to the server (there is even an issue somewhere...)

In that case, we could whitelist a property on `meta` that can be used to 'piggyback' some data back to the main process. For example `suite.meta.custom`.

Which one of these 2 approaches has your preference? `rpc` or the piggybacking approach? I will create an issue for it.

I think the 'piggybacking' strategy is simpler and covers our needs, so I would prefer that.

**Reply by sheremet-va** [reply]
The easiest to implement would be setting data on `suite.meta`. We just need to fix sending it to the main process.

**Reply by nicojs** [reply]
Thanks! I've opened https://github.com/vitest-dev/vitest/issues/3447 to keep track of the feature request. ♥
